File: convertguj.py
import csv
import json
from indic_transliteration import sanscript
from indic_transliteration.sanscript import transliterate

# Convert English digits to Gujarati
def to_gujarati_digits(num_str):
    eng = "0123456789"
    guj = "૦૧૨૩૪૫૬૭૮૯"
    return "".join(guj[eng.index(ch)] if ch in eng else ch for ch in str(num_str))

# Map flat letters to Gujarati
flat_map = {"A": "એ", "B": "બી", "C": "સી", "D": "ડી"}
def convert_flat(flat):
    parts = flat.split()
    if parts[0] in flat_map:
        return f"{flat_map[parts[0]]} {to_gujarati_digits(parts[1])}"
    return to_gujarati_digits(flat)

# Month mapping
month_map = {
    "January": "જાન્યુઆરી", "February": "ફેબ્રુઆરી", "March": "માર્ચ",
    "April": "એપ્રિલ", "May": "મે", "June": "જૂન",
    "July": "જુલાઈ", "August": "ઑગસ્ટ", "September": "સપ્ટેમ્બર",
    "October": "ઑક્ટોબર", "November": "નવેમ્બર", "December": "ડિસેમ્બર"
}

# Amount words mapping
amount_words_map = {
    5100: "પાંચ હજાર એકસો પુરા",
    5700: "પાંચ હજાર સાતસો પુરા",
    6200: "છ હજાર બે સો પુરા"
}


def transliterate_name(name):
    replacements = {
        "Hemendrabhai Panchal": "હેમેન્દ્રભાઈ પંચાલ",
        "Hemendrabhai Patel": "હેમેન્દ્રભાઈ પટેલ",
        "Pankajbhai Roy": "પંકજભાઈ રોય",
        "Jashiben Rathod": "જશીબેન રાઠોડ",
        "Akashbhai Patel": "આકાશભાઈ પટેલ",
        "Dhavalbhai Patel": "ધવલભાઈ પટેલ",
        "Niruben Bharward": "નિરુબેન ભરવર્ડ",
        "Kokilaben Rajiubhai Soni": "કોકિલાબેન રાજીઉભાઈ સોની",
        "Meraji Vanzara": "મેરાજી વાનઝારા",
        "Jitendrabhai Vanzara": "જીતેન્દ્રભાઈ વાનઝારા",
        "Ektaben V. Ramchandani": "એકટાબેન વી. રામચંદાણી",
        "Jayantibhai M. Nayak (Chiragbhai)": "જયંતિભાઈ એમ. નાયક (ચિરાગભાઈ)",
        "Ishwarbhai Sonara": "ઈશ્વરભાઈ સોનારા",
        "Hardikbhai Patel": "હાર્દિકભાઈ પટેલ",
        "Manojbhai Manaukahni": "મનોજભાઈ મનૌકહાની",
        "Bharatbhai Chauhan": "ભરતભાઈ ચૌહાણ",
        "Bharatbhai Dobariya": "ભરતભાઈ ડોબરિયા",
        "Pushpalata Devi": "પુષ્પલતા દેવી",
        "Vijay Rathod": "વિજય રાઠોડ",
        "Mukeshbhai Patel": "મુકેશભાઈ પટેલ",
        "Bharatbhai Kourani": "ભરતભાઈ કૌરાણી",
        "Hardik": "હાર્દિક",
        "Sureshbhai Joshi (Shaun)": "સુરેશભાઈ જોશી (શોન)",
        "Jigneshkumar Prajapati": "જિગ્નેશકુમાર પ્રજાપતિ",
        "Dr Rajendra Rentiyal": "ડૉ. રાજેન્દ્ર રેંટીયલ",
        "Tusharbhai Patel": "તુષારભાઈ પટેલ",
        "Gautambhai Asodiya": "ગૌરાંગભાઈ અસોદિયા",
        "Ravindrabhai Rathod": "રવિન્દ્રભાઈ રાઠોડ",
        "Ajaybhai Prajapati": "અજયભાઈ પ્રજાપતિ",
        "Pareshbhai Prajapati": "પારેશભાઈ પ્રજાપતિ",
        "Hiteshbhai Prajapti": "હિતેશભાઈ પ્રજાપ્તિ",
        "Naraynbhai Parmar": "નરાયણભાઈ પાર્મર",
        "Yogeshbhai Prajapati": "યોગેશભાઈ પ્રજાપતિ",
        "Manishaben Parmar": "મનીષાબેન પાર્મર",
        "Viralbhai Rathod": "વિરલભાઈ રાઠોડ",
        "Maulikbhai Prajapati": "મૌલિકભાઈ પ્રજાપતિ",
        "Nisargbhai Rathod": "નિસર્ગભાઈ રાઠોડ",
        "Mahendrabhai Thakkar": "મહેન્દ્રભાઈ ઠક્કર",
        "Ghanshyambhai Parekh": "ઘનશ્યામભાઈ પારેખ",
        "Parthbhai Ranpura": "પાર્થીભાઈ રાંપૂરા",
        "Minaben Jadav": "મિનાબેન જાડાવ",
        "Yogeshbhai Patel": "યોગેશભાઈ પટેલ",
        "Milanbhai Soni": "મિલનભાઈ સોની",
        "Mishalbhai Parmar": "મિશાલભાઈ પાર્મર",
        "Savanbhai Maheriya": "સાવનભાઈ મહેરિયા",
        "Atulkumar Prarjapati": "અતુલકુમાર પ્રજાપતિ",
        "Gaurangbhai Baroniya": "ગૌરાંગભાઈ બારોનિયા",
        "Harikishan Rathod": "હરિકિશન રાઠોડ",
        "Rajendrakumar Modiya": "રાજેન્દ્રકુમાર મોદિયા",
        "Nimeshkumar Priyankar": "નિમેષકુમાર પ્રિયંકર",
        "Deepakbhai Thawani": "દીપકભાઈ થાવાણી",
        "Prakashbhai Thawani": "પ્રકાશભાઈ થાવાણી",
        "R. D. Thakor": "આર. ડી. ઠાકોર",
        "Nimuben Patel": "નિમુબેન પટેલ",
        "Amrutbhai Parmar": "અમૃતભાઈ પાર્મર",
        "Mukeshbhai Bharward": "મુકેશભાઈ ભરવર્ડ"
    }

    for k, v in replacements.items():
        name = name.replace(k, v)
    return name



# Convert one row
def convert_row(row, serial):
    # Trim all keys & values
    row = {k.strip(): v.strip() for k, v in row.items()}

    # Date
    date_parts = row["Payment Date"].split()
    day = to_gujarati_digits(date_parts[0])
    month = month_map.get(date_parts[1], date_parts[1])
    year = to_gujarati_digits(date_parts[2])

    # Receipt no
    receipt_no = f"202503{serial:02d}"

    # Amount
    amount = int(row["Amount Paid"])

    # Owner Name transliteration
    # guj_name = transliterate(row["Owner Name"], sanscript.ITRANS, sanscript.GUJARATI)
    guj_name = row["Owner Name"]

    return {
        "receipt_no": to_gujarati_digits(receipt_no),
        "date": f"{day} {month} {year}",
        "name": guj_name,
        "flat_no": convert_flat(row["Flat No."]),
        "amount": to_gujarati_digits(amount),
        "amount_word": amount_words_map.get(amount, ""),
        "payment_mode": "રોકડ" if row["Mode of Payment"].lower() == "cash" else "ઓનલાઈન",
        "utr_number": to_gujarati_digits(row["Transaction Id"]) if row["Transaction Id"] else "",
        "payment_for": f"જુલાઈ થી સપ્ટેમ્બર - {year} નું મેઇન્ટેનન્સ"
    }

# Main
data = []
with open("./input.csv", newline='', encoding="utf-8") as f:
    reader = csv.DictReader(f)
    reader.fieldnames = [h.strip() for h in reader.fieldnames]
    for i, row in enumerate(reader, 1):
        data.append(convert_row(row, i))

with open("output.json", "w", encoding="utf-8") as f:
    json.dump(data, f, ensure_ascii=False, indent=2)

print("✅ Gujarati JSON created: output.json")

File: convertguj.js
import fs from 'fs';
import path from 'path';
import csv from 'csv-parser';

import { fileURLToPath } from 'url';

// Convert English digits to Gujarati
function toGujaratiDigits(numStr) {
    const eng = "0123456789";
    const guj = "૦૧૨૩૪૫૬૭૮૯";
    return String(numStr).split('').map(ch => {
        const index = eng.indexOf(ch);
        return index !== -1 ? guj[index] : ch;
    }).join('');
}

function convertFileName(flat) {
    const fileName = flat.replaceAll(" ", "_");
   
    return fileName;
}

// Map flat letters to Gujarati
const flatMap = { "A": "એ", "B": "બી", "C": "સી", "D": "ડી" };

function convertFlat(flat) {
    const parts = flat.split(' ');
    if (flatMap[parts[0]]) {
        return `${flatMap[parts[0]]}-${toGujaratiDigits(parts[1])}`;
    }
    return toGujaratiDigits(flat);
}

// Month mapping
const monthMap = {
    "January": "જાન્યુઆરી", "February": "ફેબ્રુઆરી", "March": "માર્ચ",
    "April": "એપ્રિલ", "May": "મે", "June": "જૂન",
    "July": "જુલાઈ", "August": "ઑગસ્ટ", "September": "સપ્ટેમ્બર",
    "October": "ઑક્ટોબર", "November": "નવેમ્બર", "December": "ડિસેમ્બર"
};

const monthNoMap = {
    "January": "1", "February": "2", "March": "3",
    "April": "4", "May": "5", "June": "6",
    "July": "7", "August": "8", "September": "9",
    "October": "10", "November": "11", "December": "12"
};

// Amount words mapping
const amountWordsMap = {
    5100: "પાંચ હજાર એકસો પુરા",
    5600: "પાંચ હજાર છસો પુરા",
    5700: "પાંચ હજાર સાતસો પુરા",
    6200: "છ હજાર બે સો પુરા",
};

function transliterateName(name) {
    for (const key in nameReplacements) {
        if (name.includes(key)) {
            name = name.replace(key, nameReplacements[key]);
        }
    }
    return name;
}
function getKeyfromValue(valueToFind, myObject){
    let foundKey;

    for (const key in myObject) {
        if (myObject.hasOwnProperty(key) && myObject[key] === valueToFind) {
            foundKey = key;
            break; // Exit the loop once the key is found
        }
    }
    return foundKey;
}

function getPaymentForDetail(month) {
    const monthNo = monthNoMap[month];
    const quarterMonth = getQuarterMonths(parseInt(monthNo));
    console.log("quarterMonth ",quarterMonth);
    const start = getKeyfromValue(quarterMonth.startMonth.toString(), monthNoMap);
    const end = getKeyfromValue(quarterMonth.endMonth.toString(), monthNoMap);
    console.log("end ",end)
    return {
        start: monthMap[start],
        end: monthMap[end],
    };
}

// Convert a single row
function convertRow(row, serial) {
    // Trim keys & values
    for (const key in row) {
        row[key.trim()] = row[key].trim();
    }

    // Date
    const dateParts = row["Payment Date"].split(' ');
    const day = toGujaratiDigits(dateParts[0]);
    const month = monthMap[dateParts[1]] || dateParts[1];
    const year = toGujaratiDigits(dateParts[2]);

    const payment_for = getPaymentForDetail(dateParts[1]);
    // Receipt no
    const receiptNo = `202503${String(serial).padStart(2, '0')}`;
    console.log("payment_for ",payment_for)
    console.log("month ",month)
    // Amount
    const amount = parseInt(row["Amount Paid"]);

    // Owner Name transliteration
    // Note: The original code has a comment for a library that's not available in Node.js.
    // The provided manual map is used instead.
    const gujName = row["Owner Name"];

    return {
        "receipt_no": toGujaratiDigits(receiptNo),
        "date": `${day} ${month} ${year}`,
        "name": gujName,
        "file_name": convertFileName(row["Flat No."]),
        "flat_no": convertFlat(row["Flat No."]),
        "amount": toGujaratiDigits(amount),
        "amount_word": amountWordsMap[amount] || "",
        "payment_mode": row["Mode of Payment"].toLowerCase() === "cash" ? "રોકડ" : "ઓનલાઈન",
        "utr_number": row["Transaction Id"] ? toGujaratiDigits(row["Transaction Id"]) : "",
        "payment_for": `${payment_for.start} થી ${payment_for.end} - ${year} નું મેઇન્ટેનન્સ`
    };
}


function createDateFromDMY(dateString) {
    const parts = dateString.split('-');
    // Parts will be [day, month, year]
    const day = parseInt(parts[0], 10);
    const month = parseInt(parts[1], 10) - 1; // Month is 0-indexed
    const year = parseInt(parts[2], 10);

    return new Date(year, month, day);
}

function getQuarterMonths(month) {
    const quarter = Math.floor(month / 3);
    console.log("quarter ",quarter);
    const startMonthIndex = quarter * 3;
    const endMonthIndex = startMonthIndex + 2;

    // Return 1-indexed month numbers for user clarity
    return {
        startMonth: startMonthIndex + 1,
        endMonth: endMonthIndex + 1
    };
}

// Main function to process the CSV file
async function main() {
    const results = [];
    let serial = 1;


    // Convert the current module's URL to a file path
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);
    fs.createReadStream(path.join(__dirname, 'input.csv'))
        .pipe(csv({
            mapHeaders: ({ header }) => header.trim()
        }))
        .on('data', (data) => {
            results.push(convertRow(data, serial));
            serial++;
        })
        .on('end', () => {
            fs.writeFileSync('output.json', JSON.stringify(results, null, 2), 'utf-8');
            console.log("✅ Gujarati JSON created: output.json");
        })
        .on('error', (err) => {
            console.error("❌ An error occurred:", err);
        });
}

main();